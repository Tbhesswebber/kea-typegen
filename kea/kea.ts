// This will be integrated into Kea soon... It's just faster to iterate here.

import { kea as regularKea, useValues as useRegularValues } from 'kea'

type BlankLogic = {
    actions?: any
    actionCreators?: any
    actionTypes?: any
    reducers?: any
    selectors?: any
    listeners?: any
    values?: any
    __selectorTypeHelp?: any
}

type ReducerActions<LogicType extends BlankLogic, ReducerType> = {
    [K in keyof LogicType['actions']]?: (
        state: any,
        payload: ReturnType<LogicType['actions'][K]>['payload'],
    ) => ReducerType
}
type ReducerDefinitions<LogicType extends BlankLogic> = {
    [K in keyof LogicType['reducers']]?:
        | [ReturnType<LogicType['reducers'][K]>, ReducerActions<LogicType, ReturnType<LogicType['reducers'][K]>>]
        | ReducerActions<LogicType, ReturnType<LogicType['reducers'][K]>>
}
type Selector = (state?: any, props?: any) => any

type SelectorDefinition<Selectors, S extends Selector, SelectorFunction extends any> = [
    (
        s: Selectors,
    ) =>
        | []
        | [Selector]
        | [Selector, Selector]
        | [Selector, Selector, Selector]
        | [Selector, Selector, Selector, Selector]
        | [Selector, Selector, Selector, Selector, Selector]
        | [Selector, Selector, Selector, Selector, Selector, Selector]
        | [Selector, Selector, Selector, Selector, Selector, Selector, Selector]
        | [Selector, Selector, Selector, Selector, Selector, Selector, Selector, Selector]
        | [Selector, Selector, Selector, Selector, Selector, Selector, Selector, Selector, Selector]
        | [Selector, Selector, Selector, Selector, Selector, Selector, Selector, Selector, Selector, Selector]
        | [Selector, Selector, Selector, Selector, Selector, Selector, Selector, Selector, Selector, Selector, Selector],
    SelectorFunction,
]

type SelectorDefinitions<LogicType extends BlankLogic> = {
    [K in keyof LogicType['selectors']]?: SelectorDefinition<
        LogicType['selectors'],
        LogicType['selectors'][K],
        LogicType['__selectorTypeHelp'][K]
    >
}

type ListenerDefinitions<LogicType extends BlankLogic> = {
    [K in keyof LogicType['actions']]?:
        | ((
              payload: ReturnType<LogicType['actions'][K]>['payload'],
              breakpoint: (() => void) | ((ms: number) => Promise<void>),
              action: ReturnType<LogicType['actions'][K]>,
              previousState: any,
          ) => void | Promise<void>)
        | (() => void | Promise<void>)
}
type LogicInput<LogicType> = {
    path?: any
    constants?: any
    actions?: (logic: LogicType) => any | any
    reducers?: ReducerDefinitions<LogicType> | ((logic: LogicType) => ReducerDefinitions<LogicType>)
    events?: (logic: LogicType) => any | any
    selectors?: SelectorDefinitions<LogicType> | ((logic: LogicType) => SelectorDefinitions<LogicType>)
    listeners?: ListenerDefinitions<LogicType> | ((logic: LogicType) => ListenerDefinitions<LogicType>)
    loaders?: any
}

export function kea<LogicType>(input: LogicInput<LogicType>): LogicType {
    return regularKea(input)
}
export function useValues<L extends BlankLogic>(logic: L): L['values'] {
    return useRegularValues(logic)
}
