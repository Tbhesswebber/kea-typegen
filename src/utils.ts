import * as ts from 'typescript'
import * as path from 'path'
import { cloneNode } from '@wessberg/ts-clone-node'
import { visitProgram } from './visit/visit'
import { parsedLogicToTypeString } from './print/print'
import { AppOptions, NameType, ParsedLogic } from './types'
import { factory, NodeBuilderFlags, SyntaxKind } from 'typescript'

export function logicSourceToLogicType(logicSource: string, appOptions?: AppOptions) {
    const program = programFromSource(logicSource)
    const [parsedLogic] = visitProgram(program)
    return parsedLogicToTypeString(parsedLogic, appOptions)
}

export function sourceToSourceFile(sourceCode: string, filename: string = 'logic.ts') {
    return ts.createSourceFile(filename, sourceCode, ts.ScriptTarget.ES5, true)
}

export function programFromSource(sourceCode: string) {
    const options = {}
    const compilerHost = ts.createCompilerHost(options)
    compilerHost.getSourceFile = (fileName) => (fileName === 'logic.ts' ? sourceToSourceFile(sourceCode) : undefined)
    return ts.createProgram(['logic.ts'], options, compilerHost)
}

export function isKeaCall(node: ts.Node, checker: ts.TypeChecker) {
    if (!ts.isIdentifier(node)) {
        return false
    }

    if (!node.parent || !ts.isCallExpression(node.parent)) {
        return false
    }

    const symbol = checker.getSymbolAtLocation(node)
    if (!symbol || symbol.getName() !== 'kea') {
        return false
    }

    const input = node.parent.arguments[0]

    if (!input || !ts.isObjectLiteralExpression(input)) {
        return false
    }

    return true
}

export function getTypeNodeForNode(node: ts.Node, checker: ts.TypeChecker): ts.TypeNode {
    let typeNode
    if (node) {
        if (ts.isAsExpression(node)) {
            typeNode = node.type
            if (ts.isParenthesizedTypeNode(typeNode)) {
                typeNode = typeNode.type
            }
        } else if (node?.kind === SyntaxKind.TrueKeyword || node?.kind === SyntaxKind.FalseKeyword) {
            typeNode = factory.createKeywordTypeNode(SyntaxKind.BooleanKeyword)
        } else if (ts.isStringLiteralLike(node)) {
            typeNode = factory.createKeywordTypeNode(SyntaxKind.StringKeyword)
        } else if (ts.isNumericLiteral(node)) {
            typeNode = factory.createKeywordTypeNode(SyntaxKind.NumberKeyword)
        } else if (ts.isArrayLiteralExpression(node) && node.elements.length === 0) {
            typeNode = factory.createArrayTypeNode(factory.createKeywordTypeNode(SyntaxKind.AnyKeyword))
        } else {
            typeNode = checker.typeToTypeNode(checker.getTypeAtLocation(node), undefined, undefined)
        }
    } else {
        typeNode = factory.createKeywordTypeNode(SyntaxKind.AnyKeyword)
    }
    return typeNode
}

export function getAndGatherTypeNodeForDefaultValue(
    defaultValue: ts.Node,
    checker: ts.TypeChecker,
    parsedLogic: ParsedLogic,
): ts.TypeNode {
    const typeNode = getTypeNodeForNode(defaultValue, checker)
    gatherImports(typeNode, checker, parsedLogic)
    return cloneNode(typeNode)
}

export function getParameterDeclaration(param: ts.ParameterDeclaration) {
    return factory.createParameterDeclaration(
        undefined,
        undefined,
        undefined,
        factory.createIdentifier(param.name.getText()),
        param.initializer || param.questionToken ? factory.createToken(SyntaxKind.QuestionToken) : undefined,
        cloneNode(param.type || factory.createKeywordTypeNode(SyntaxKind.AnyKeyword)),
        undefined,
    )
}

export const toSpaces = (key) => key.replace(/(?:^|\.?)([A-Z])/g, (x, y) => ' ' + y.toLowerCase()).replace(/^ /, '')

export function getActionTypeCreator(parsedLogic: ParsedLogic) {
    return function (actionName) {
        return `${toSpaces(actionName)} (${parsedLogic.pathString})`
    }
}

export function cleanDuplicateAnyNodes(reducers: NameType[]): NameType[] {
    let newReducers = {}

    for (const reducer of reducers) {
        // if (reducer.name === 'featureFlag') {
        //     debugger
        // }
        if (!newReducers[reducer.name] || !isAnyUnknown(reducer.typeNode)) {
            newReducers[reducer.name] = reducer
        }
    }

    return Object.values(newReducers)
}

export function extractImportedActions(
    actionObjects: ts.Expression | ts.ObjectLiteralExpression,
    checker: ts.TypeChecker,
    parsedLogic: ParsedLogic,
) {
    let extraActions = {}

    if (ts.isObjectLiteralExpression(actionObjects)) {
        // actionObjects =  { [githubLogic.actionTypes.setRepositories]: () => ... }
        for (const property of actionObjects.properties) {
            // property.name ==> [githubLogic.actionTypes.setRepositories]
            if (ts.isComputedPropertyName(property.name)) {
                let propertyExpression = property.name.expression

                if (ts.isPropertyAccessExpression(propertyExpression)) {
                    const { name, expression } = propertyExpression
                    const actionName = name.escapedText

                    const nameSymbol = checker.getSymbolAtLocation(property.name)
                    let actionType = nameSymbol.escapedName as string

                    if (ts.isPropertyAccessExpression(expression)) {
                        // expression.expression ==> githubLogic.actionTypes
                        // expression.name ==> setRepositories

                        const logicForAction = ts.isCallExpression(expression.expression)
                            ? expression.expression.expression
                            : expression.expression
                        const symbol = checker.getSymbolAtLocation(logicForAction)
                        const symbolType = checker.getTypeOfSymbolAtLocation(symbol, logicForAction)

                        const actionCreatorsProperty = symbolType
                            .getProperties()
                            .find((p) => p.escapedName === 'actionCreators')
                        const actionCreators = actionCreatorsProperty?.valueDeclaration

                        if (
                            actionCreators &&
                            ts.isPropertySignature(actionCreators) &&
                            ts.isTypeLiteralNode(actionCreators.type)
                        ) {
                            const actionCreator = actionCreators.type.members.find(
                                (m) => (m.name as ts.Identifier)?.escapedText === actionName,
                            )

                            if (
                                actionCreator &&
                                ts.isPropertySignature(actionCreator) &&
                                ts.isFunctionTypeNode(actionCreator.type)
                            ) {
                                const type = checker.getTypeFromTypeNode(actionCreator.type)
                                const signature = type.getCallSignatures()[0]
                                const sigReturnType = signature.getReturnType()
                                const sigReturnTypeNode = checker.typeToTypeNode(
                                    sigReturnType,
                                    undefined,
                                    NodeBuilderFlags.NoTruncation,
                                )
                                gatherImports(sigReturnTypeNode, checker, parsedLogic)

                                if (actionType === '__computed') {
                                    if (ts.isTypeNode(actionCreator.type) && ts.isTypeNode(actionCreator.type.type)) {
                                        const m = (actionCreator.type.type as any).members?.find(
                                            (m) => m.name?.getText() === 'type',
                                        )
                                        if (ts.isPropertySignature(m) && ts.isLiteralTypeNode(m.type)) {
                                            const str = m.type.getText()
                                            if (str) {
                                                actionType = str.substring(1, str.length - 1)
                                            }
                                        }
                                    }
                                }

                                extraActions[actionType] = cloneNode(actionCreator.type) //payload
                            }
                        }
                    }
                }
            }
        }
    }
    return extraActions
}

export function getLogicPathString(appOptions: AppOptions, fileName: string) {
    let cwd = process.cwd()
    if (appOptions?.rootPath) {
        cwd = path.resolve(cwd, appOptions.rootPath)
    }
    const pathString = path
        .relative(cwd, fileName)
        .replace(/^.\//, '')
        .replace(/\.[jt]sx?$/, '')
        .replace(/\//g, '.')

    return pathString
}

export function getFilenamesForSymbol(symbol: ts.Symbol): string[] | undefined {
    return (symbol?.declarations || [])
        .map((d) => d.getSourceFile().fileName)
        .filter((str) => !str.includes('/node_modules/typescript/lib/lib'))
}

/** gathers onto parsedLogic the TypeReference nodes that are declared in a different sourceFile */
export function gatherImports(input: ts.Node, checker: ts.TypeChecker, parsedLogic: ParsedLogic) {
    if (!input) {
        return
    }
    function getImports(requestedNode) {
        let node = requestedNode
        if (ts.isTypeReferenceNode(node)) {
            let typeRootName: string | undefined
            if (node.typeName?.kind === SyntaxKind.FirstNode) {
                try {
                    typeRootName = node.typeName.getFirstToken().getText()
                } catch (e) {
                    typeRootName = (node.typeName.left as any)?.escapedText
                }
            }

            const symbol = checker.getSymbolAtLocation(node.typeName) || (node.typeName as any).symbol
            if (symbol) {
                storeExtractedSymbol(symbol, checker, parsedLogic, typeRootName)
            }
        }
        ts.forEachChild(requestedNode, getImports)
    }
    getImports(input)
}

export function storeExtractedSymbol(
    symbol: ts.Symbol,
    checker: ts.TypeChecker,
    parsedLogic: ParsedLogic,
    typeRootName?: string,
) {
    const declaration = symbol.getDeclarations()?.[0]

    if (declaration && ts.isImportSpecifier(declaration)) {
        const importFilename = getFilenameForImportSpecifier(declaration, checker)
        if (importFilename) {
            addTypeImport(parsedLogic, importFilename, typeRootName || declaration.getText())
        } else {
            parsedLogic.typeReferencesInLogicInput.add(typeRootName || declaration.getText())
        }
        return
    }

    const files = getFilenamesForSymbol(symbol)
    if (files[0]) {
        // same file, add to logicType<...>
        if (
            ts.isTypeAliasDeclaration(declaration) ||
            ts.isInterfaceDeclaration(declaration) ||
            ts.isEnumDeclaration(declaration) ||
            ts.isClassDeclaration(declaration)
        ) {
            if (files[0] === parsedLogic.fileName) {
                parsedLogic.typeReferencesInLogicInput.add(typeRootName || declaration.name.getText())
            } else {
                // but is it exported?
                addTypeImport(parsedLogic, files[0], typeRootName || declaration.name.getText())
            }
        }
    }
}

export function getFilenameForImportDeclaration(checker: ts.TypeChecker, importNode: ts.ImportDeclaration): string {
    const moduleSymbol = checker.getSymbolAtLocation(importNode.moduleSpecifier)
    const otherSourceFile = moduleSymbol?.getDeclarations()[0].getSourceFile()
    if (otherSourceFile) {
        return otherSourceFile.fileName || importNode.moduleSpecifier.getText()
    }
}

export function getFilenameForImportSpecifier(declaration: ts.ImportSpecifier, checker: ts.TypeChecker): string | void {
    let importNode: ts.Node = declaration
    while (importNode && !ts.isImportDeclaration(importNode)) {
        importNode = importNode.parent
    }
    if (ts.isImportDeclaration(importNode)) {
        return getFilenameForImportDeclaration(checker, importNode)
    }
}

function addTypeImport(parsedLogic: ParsedLogic, file: string, typeName: string) {
    if (!parsedLogic.typeReferencesToImportFromFiles[file]) {
        parsedLogic.typeReferencesToImportFromFiles[file] = new Set()
    }
    parsedLogic.typeReferencesToImportFromFiles[file].add(typeName.split('.')[0])
}

export function arrayContainsSet(array: string[], setToContain: Set<string>): boolean {
    const arraySet = new Set(array)
    for (const str of setToContain) {
        if (!arraySet.has(str)) {
            return false
        }
    }
    return true
}

export function unPromisify(node: ts.Node): ts.Node {
    if (ts.isTypeReferenceNode(node) && (node.typeName as any)?.escapedText === 'Promise') {
        return node.typeArguments?.[0]
    }
    return node
}

export function isAnyUnknown(node?: ts.Node): boolean {
    if (!node) {
        return true
    }
    const unPromised = unPromisify(node)
    return (
        !unPromised ||
        unPromised.kind === SyntaxKind.AnyKeyword ||
        unPromised.kind === SyntaxKind.UnknownKeyword ||
        (ts.isTypeLiteralNode(unPromised) && unPromised.members.length === 0)
    )
}
