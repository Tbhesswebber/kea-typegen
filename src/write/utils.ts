import { ParsedLogic } from '../types'
import { visit, types } from 'recast'

export const t = types.namedTypes
export const b = types.builders

export function isKeaCall(path: any): boolean {
    const stmt = path.node
    return (
        t.Identifier.check(stmt.callee) &&
        stmt.callee.name === 'kea' &&
        stmt.arguments[0] &&
        (t.ObjectExpression.check(stmt.arguments[0]) || t.ArrayExpression.check(stmt.arguments[0])) &&
        path.parentPath &&
        t.VariableDeclarator.check(path.parentPath.value) &&
        t.Identifier.check(path.parentPath.value.id)
    )
}

export function visitAllKeaCalls(
    ast: any,
    parsedLogics: ParsedLogic[],
    filename: string,
    callback: (args: { parsedLogic: ParsedLogic; logicName: string; path: any }) => any,
): void {
    visit(ast, {
        visitCallExpression: function (path) {
            if (!isKeaCall(path)) {
                this.traverse(path)
            }
            const logicName = path.parentPath.value.id.name
            if (!logicName) {
                console.warn(
                    `[KEA-TYPEGEN] Can not add path to logic in "${filename}:${path.node.loc.start}" because it's not stored as a variable.`,
                )
                return
            }
            const parsedLogic = parsedLogics.find((l) => l.logicName === logicName)
            if (!parsedLogic) {
                console.error(
                    `[KEA-TYPEGEN] While trying to add a path, could not find logicName "${logicName}" in the list of logicNames (${Object.keys(
                        parsedLogics.map((l) => l.logicName),
                    ).join(', ')}) in the file: ${filename}`,
                )
                return
            }
            callback.bind(this)({ logicName, parsedLogic, path })
            return false
        },
    })
}
