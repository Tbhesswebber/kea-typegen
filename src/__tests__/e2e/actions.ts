import { logicSourceToLogicType } from '../../utils'

test('actions - with a function', () => {
    const logicSource = `
        import { kea } from 'kea'
        
        const logic = kea({
            actions: () => ({
                updateName: (name: string) => ({ name }),
                updateOtherName: (otherName: string) => ({ otherName }),
            })
        })
    `
    expect(logicSourceToLogicType(logicSource)).toEqual(
        `
export interface logicType {
    actionCreators: {
        updateName: (name: string) => ({
            type: string;
            payload: { name: string; };
        });
        updateOtherName: (otherName: string) => ({
            type: string;
            payload: { otherName: string; };
        });
    };
    actions: {
        updateName: (name: string) => ({
            type: string;
            payload: { name: string; };
        });
        updateOtherName: (otherName: string) => ({
            type: string;
            payload: { otherName: string; };
        });
    };
    reducer: (state: any, action: () => any, fullState: any) => {};
    reducers: {};
    selector: (state: any) => {};
    selectors: {};
    values: {};
}`.trim(),
    )
})

test('actions - as an object', () => {
    const logicSource = `
        import { kea } from 'kea'
        
        const logic = kea({
            actions: {
                updateName: (name: string) => ({ name }),
                updateOtherName: (otherName: string) => ({ otherName }),
            }
        })
    `
    expect(logicSourceToLogicType(logicSource)).toEqual(
        `
export interface logicType {
    actionCreators: {
        updateName: (name: string) => ({
            type: string;
            payload: { name: string; };
        });
        updateOtherName: (otherName: string) => ({
            type: string;
            payload: { otherName: string; };
        });
    };
    actions: {
        updateName: (name: string) => ({
            type: string;
            payload: { name: string; };
        });
        updateOtherName: (otherName: string) => ({
            type: string;
            payload: { otherName: string; };
        });
    };
    reducer: (state: any, action: () => any, fullState: any) => {};
    reducers: {};
    selector: (state: any) => {};
    selectors: {};
    values: {};
}`.trim(),
    )
})

test('actions - as a function returning a object', () => {
    const logicSource = `
        import { kea } from 'kea'
        
        const logic = kea({
            actions: function () {
                return {
                    updateName: (name: string) => ({ name }),
                    updateOtherName: (otherName: string) => ({ otherName }),
                }
            }
        })
    `
    expect(logicSourceToLogicType(logicSource)).toEqual(
        `
export interface logicType {
    actionCreators: {
        updateName: (name: string) => ({
            type: string;
            payload: { name: string; };
        });
        updateOtherName: (otherName: string) => ({
            type: string;
            payload: { otherName: string; };
        });
    };
    actions: {
        updateName: (name: string) => ({
            type: string;
            payload: { name: string; };
        });
        updateOtherName: (otherName: string) => ({
            type: string;
            payload: { otherName: string; };
        });
    };
    reducer: (state: any, action: () => any, fullState: any) => {};
    reducers: {};
    selector: (state: any) => {};
    selectors: {};
    values: {};
}`.trim(),
    )
})

test('actions - with random values instead of functions', () => {
    const logicSource = `
        import { kea } from 'kea'
        
        const logic = kea({
            actions: {
                updateName: (name: string) => ({ name }),
                withDefaultValue: (name = "john") => ({ name }),
                withDefaultValueAndType: (name: string = "john") => ({ name }),
                withBool: true,
                withRandomPayload: { bla: 123 },
            }
        })
    `
    expect(logicSourceToLogicType(logicSource)).toEqual(
        `
export interface logicType {
    actionCreators: {
        updateName: (name: string) => ({
            type: string;
            payload: { name: string; };
        });
        withDefaultValue: (name: any) => ({
            type: string;
            payload: { name: string; };
        });
        withDefaultValueAndType: (name: string) => ({
            type: string;
            payload: { name: string; };
        });
        withBool: () => ({
            type: string;
            payload: {
                value: boolean;
            };
        });
        withRandomPayload: () => ({
            type: string;
            payload: {
                value: { bla: number; };
            };
        });
    };
    actions: {
        updateName: (name: string) => ({
            type: string;
            payload: { name: string; };
        });
        withDefaultValue: (name: any) => ({
            type: string;
            payload: { name: string; };
        });
        withDefaultValueAndType: (name: string) => ({
            type: string;
            payload: { name: string; };
        });
        withBool: () => ({
            type: string;
            payload: {
                value: boolean;
            };
        });
        withRandomPayload: () => ({
            type: string;
            payload: {
                value: { bla: number; };
            };
        });
    };
    reducer: (state: any, action: () => any, fullState: any) => {};
    reducers: {};
    selector: (state: any) => {};
    selectors: {};
    values: {};
}`.trim(),
    )
})
