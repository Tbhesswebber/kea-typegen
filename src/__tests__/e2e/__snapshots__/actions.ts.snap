// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions - as a function returning a object 1`] = `
"export interface logicType {
    key: any;
    actionCreators: {
        updateName: (name: string) => ({
            type: \\"update name (logic)\\";
            payload: { name: string; };
        });
        updateOtherName: (otherName: string) => ({
            type: \\"update other name (logic)\\";
            payload: { otherName: string; };
        });
    };
    actionKeys: Record<string, any>;
    actionTypes: Record<string, any>;
    actions: {
        updateName: (name: string) => ({
            type: \\"update name (logic)\\";
            payload: { name: string; };
        });
        updateOtherName: (otherName: string) => ({
            type: \\"update other name (logic)\\";
            payload: { otherName: string; };
        });
    };
    cache: Record<string, any>;
    connections: any;
    constants: any;
    defaults: any;
    events: any;
    path: [\\"logic\\"];
    pathString: \\"logic\\";
    propTypes: any;
    props: Record<string, any>;
    reducer: (state: any, action: () => any, fullState: any) => {};
    reducerOptions: any;
    reducers: {};
    selector: (state: any) => {};
    selectors: {};
    values: {};
    _isKea: true;
}"
`;

exports[`actions - as an object 1`] = `
"export interface logicType {
    key: any;
    actionCreators: {
        updateName: (name: string) => ({
            type: \\"update name (logic)\\";
            payload: { name: string; };
        });
        updateOtherName: (otherName: string) => ({
            type: \\"update other name (logic)\\";
            payload: { otherName: string; };
        });
    };
    actionKeys: Record<string, any>;
    actionTypes: Record<string, any>;
    actions: {
        updateName: (name: string) => ({
            type: \\"update name (logic)\\";
            payload: { name: string; };
        });
        updateOtherName: (otherName: string) => ({
            type: \\"update other name (logic)\\";
            payload: { otherName: string; };
        });
    };
    cache: Record<string, any>;
    connections: any;
    constants: any;
    defaults: any;
    events: any;
    path: [\\"logic\\"];
    pathString: \\"logic\\";
    propTypes: any;
    props: Record<string, any>;
    reducer: (state: any, action: () => any, fullState: any) => {};
    reducerOptions: any;
    reducers: {};
    selector: (state: any) => {};
    selectors: {};
    values: {};
    _isKea: true;
}"
`;

exports[`actions - with a function 1`] = `
"export interface logicType {
    key: any;
    actionCreators: {
        updateName: (name: string) => ({
            type: \\"update name (logic)\\";
            payload: { name: string; };
        });
        updateOtherName: (otherName: string) => ({
            type: \\"update other name (logic)\\";
            payload: { otherName: string; };
        });
    };
    actionKeys: Record<string, any>;
    actionTypes: Record<string, any>;
    actions: {
        updateName: (name: string) => ({
            type: \\"update name (logic)\\";
            payload: { name: string; };
        });
        updateOtherName: (otherName: string) => ({
            type: \\"update other name (logic)\\";
            payload: { otherName: string; };
        });
    };
    cache: Record<string, any>;
    connections: any;
    constants: any;
    defaults: any;
    events: any;
    path: [\\"logic\\"];
    pathString: \\"logic\\";
    propTypes: any;
    props: Record<string, any>;
    reducer: (state: any, action: () => any, fullState: any) => {};
    reducerOptions: any;
    reducers: {};
    selector: (state: any) => {};
    selectors: {};
    values: {};
    _isKea: true;
}"
`;

exports[`actions - with random values instead of functions 1`] = `
"export interface logicType {
    key: any;
    actionCreators: {
        updateName: (name?: string) => ({
            type: \\"update name (logic)\\";
            payload: { name: string; };
        });
        withDefaultValue: (name?: any) => ({
            type: \\"with default value (logic)\\";
            payload: { name: string; };
        });
        withDefaultValueAndType: (name?: string) => ({
            type: \\"with default value and type (logic)\\";
            payload: { name: string; };
        });
        withBool: () => ({
            type: \\"with bool (logic)\\";
            payload: {
                value: boolean;
            };
        });
        withRandomPayload: () => ({
            type: \\"with random payload (logic)\\";
            payload: {
                value: { bla: number; };
            };
        });
    };
    actionKeys: Record<string, any>;
    actionTypes: Record<string, any>;
    actions: {
        updateName: (name?: string) => ({
            type: \\"update name (logic)\\";
            payload: { name: string; };
        });
        withDefaultValue: (name?: any) => ({
            type: \\"with default value (logic)\\";
            payload: { name: string; };
        });
        withDefaultValueAndType: (name?: string) => ({
            type: \\"with default value and type (logic)\\";
            payload: { name: string; };
        });
        withBool: () => ({
            type: \\"with bool (logic)\\";
            payload: {
                value: boolean;
            };
        });
        withRandomPayload: () => ({
            type: \\"with random payload (logic)\\";
            payload: {
                value: { bla: number; };
            };
        });
    };
    cache: Record<string, any>;
    connections: any;
    constants: any;
    defaults: any;
    events: any;
    path: [\\"logic\\"];
    pathString: \\"logic\\";
    propTypes: any;
    props: Record<string, any>;
    reducer: (state: any, action: () => any, fullState: any) => {};
    reducerOptions: any;
    reducers: {};
    selector: (state: any) => {};
    selectors: {};
    values: {};
    _isKea: true;
}"
`;
